### PR TITLE
feat: remove filters with null constants

### LINKED ISSUE #8689: Remove filters for NULL constants (in addition to `false`)
### Is your feature request related to a problem or challenge?

DataFusion tries to avoid doing work when at all possible to improve query performance

Part of this work is to determine when filters can never be true and avoid doing work

For example

```sql
DataFusion CLI v34.0.0
❯ create table t(x int) as values (1), (2), (3);
0 rows in set. Query took 0.003 seconds.
```

When DataFusion sees a filter that can't be true it skips even scanning the data
```
❯ explain select x from t where false;
+---------------+---------------+
| plan_type     | plan          |
+---------------+---------------+
| logical_plan  | EmptyRelation |
| physical_plan | EmptyExec     |
|               |               |
+---------------+---------------+
2 rows in set. Query took 0.001 seconds.
```

However, it currently does not skip scanning if the filter is NULL (which also can't be true). Note the `MemoryExec` is still present:

```
❯ explain select x from t where null::bool;
+---------------+---------------------------------------------------+
| plan_type     | plan                                              |
+---------------+---------------------------------------------------+
| logical_plan  | Filter: Boolean(NULL)                             |
|               |   TableScan: t projection=[x]                     |
| physical_plan | CoalesceBatchesExec: target_batch_size=8192       |
|               |   FilterExec: NULL                                |
|               |     MemoryExec: partitions=1, partition_sizes=[1] |
|               |                                                   |
+---------------+---------------------------------------------------+
```

### Describe the solution you'd like

I would like to avoid scanning when the filter evaluates to NULL in addition to `false` (the second example above should not have a `MemoryExec` in it)

### Describe alternatives you've considered

_No response_

### Additional context

I think this is a good first issue that should be relatively simple to implement and would be a good introduction to DataFusion

### PR DESCRIPTION
**Pull request by asimsedhain**
## Which issue does this PR close?

Closes #8689

## Are these changes tested?

```bash
DataFusion CLI v34.0.0
❯ create table t(x int) as values (1), (2), (3);
0 rows in set. Query took 0.031 seconds.

❯ explain select x from t where null::bool;
+---------------+---------------+
| plan_type     | plan          |
+---------------+---------------+
| logical_plan  | EmptyRelation |
| physical_plan | EmptyExec     |
|               |               |
+---------------+---------------+
2 rows in set. Query took 0.005 seconds.
```

### CODE REVIEW

File: datafusion/optimizer/src/eliminate_filter.rs [hunk 1]
```diff
@@ -62,6 +62,14 @@ impl OptimizerRule for EliminateFilter {
                     }))),
                 }
             }
+            LogicalPlan::Filter(Filter {
+                predicate: Expr::Literal(ScalarValue::Boolean(None)),
+                input,
+                ..
+            }) => Ok(Some(LogicalPlan::EmptyRelation(EmptyRelation {
```
Comment (jonahgao) [inline datafusion/optimizer/src/eliminate_filter.rs:69]: How about merging it into the previous `false` match arm like this 🤔?
```rust
fn try_optimize(
        &self,
        plan: &LogicalPlan,
        _config: &dyn OptimizerConfig,
    ) -> Result<Option<LogicalPlan>> {
        match plan {
            LogicalPlan::Filter(Filter {
                predicate: Expr::Literal(ScalarValue::Boolean(v)),
                input,
                ..
            }) => {
                match *v {
                    // input also can be filter, apply again
                    Some(true) => Ok(Some(
                        self.try_optimize(input, _config)?
                            .unwrap_or_else(|| input.as_ref().clone()),
                    )),
                    Some(false) | None => {
                        Ok(Some(LogicalPlan::EmptyRelation(EmptyRelation {
                            produce_one_row: false,
                            schema: input.schema().clone(),
                        })))
                    }
                }
            }
            _ => Ok(None),
        }
    }
```

Comment (asimsedhain) [reply datafusion/optimizer/src/eliminate_filter.rs]: Updated

**Review by jonahgao — APPROVED** [review]
Looks good to me!

**Review by alamb — APPROVED** [review]
Looks great -- thank you @asimsedhain  and @jonahgao  for the review

**Pull request merged**